src/combinatorics/simple_graph/connectivity.lean

```diff
@@ -94,6 +95,35 @@ variables {G}
 @[pattern] abbreviation cons' (u v w : V) (h : G.adj u v) (p : G.walk v w) : G.walk u w :=
 walk.cons h p
 
+/-- Change the endpoints of a walk using equalities. This is helpful for relaxing
+definitional equality constraints and to be able to state otherwise difficult-to-state
+lemmas. While this is a simple wrapper around `eq.rec`, it gives a canonical way to write it.
+
+The simp-normal form is for the `copy` to be pushed outward. That way calculations can
+occur within the "copy context." -/
+protected def copy {u v u' v'} (p : G.walk u v) (hu : u = u') (hv : v = v') : G.walk u' v' :=
+eq.rec (eq.rec p hv) hu
+
+@[simp] lemma copy_rfl_rfl {u v} (p : G.walk u v) (hu : u = u) (hv : v = v) :
+  p.copy hu hv = p := rfl
+
+@[simp] lemma copy_copy {u v u' v' u'' v''} (p : G.walk u v)
+  (hu : u = u') (hv : v = v') (hu' : u' = u'') (hv' : v' = v'') :
+  (p.copy hu hv).copy hu' hv' = p.copy (hu.trans hu') (hv.trans hv') :=
+by { subst_vars, refl }
+
+@[simp] lemma copy_nil {u u'} (hu : u = u') : (walk.nil : G.walk u u).copy hu hu = walk.nil :=
+by { subst_vars, refl }
+
+lemma copy_cons {u v w u' w'} (h : G.adj u v) (p : G.walk v w) (hu : u = u') (hw : w = w') :
+  (walk.cons h p).copy hu hw = walk.cons (by rwa ← hu) (p.copy rfl hw) :=
```

Comment:
This and the below don't seem to follow the pattern of pushing `copy` outwards, or am I missing something?

Reply:
This lemma is because sometimes you need to make something more complicated first, to do a rewrite then re-simplify. The converse to `cons_copy` is too specialized.

Reply:
Agreed, but then doesn't this break the description earlier about simp-normal forms involving `copy`?

Reply:
This one's not `simp` lemma though, unless I'm missing your point here.

Reply:
Yeah but the next one is!

Reply:
Could you be more specific here? The next lemma is a `simp` lemma, sure, but I'm not sure why this non-`simp` lemma breaks the rule I described about `simp` normal forms.